### DISCUSSION TITLE
On the problem of slint memory leak

### DISCUSSION THREAD

**Discussion by pcevent** [slint-ui/slint#2750] [Q&A]
Hello!
I made a small demo (using a timer to continuously increment a value on the interface). During operation, the memory is constantly increasing. Is it a bug or a cache caused? If it is caused by caching, how should it be turned off?

appwindow.slint
```
export component AppWindow inherits Window {
    property<int> counter: 42;
    callback inc();
    inc() => {
        counter += 1;
    }
    Text {
        text: "Counter: \{root.counter}";
    }
    width: 200px;
    height: 100px;
}
```

main.rust
```rust
slint::include_modules!();

use slint::Timer;

fn main() -> Result<(), slint::PlatformError> {
    let ui = AppWindow::new()?;

    let timer = Timer::default();
    let uh = ui.as_weak();
    timer.start(slint::TimerMode::Repeated, std::time::Duration::from_millis(20), move || {
        let ui = uh.unwrap();
        ui.invoke_inc();
    });

    ui.run()
}
```

**Comment by tronical** [comment]
Hi! Thanks for the report. What platform are you running this on? Do you know what backend you're using? (winit or Qt)

**Comment by pcevent** [comment]
I am running on MacOS and using winit.

**Comment by ogoffart** [comment]
What's your platform (opereting system)  and do you know what graphical backend you are using?  (Qt, Femtovg or skia?)

The memory should increase too much.

Maybe vallgrind's massif or [heaptrack](https://github.com/KDE/heaptrack) tools can help to find out where the memory is spent.

We do have some cache. For example the image cache https://github.com/slint-ui/slint/blob/757a744123c9ad545837e04483aa095d3e6e0150/internal/core/graphics/image/cache.rs#L46 which van be 5M big.
But that example shouldn't be using the image cache.
There is another cache in femtovg for fonts. but it shouldn't increase since that's always the same font.
there might be some other caches.

**Comment by pcevent** [comment]
Above is the complete code。

<img width="407" alt="截屏2023-05-17 22 32 02" src="https://github.com/slint-ui/slint/assets/33011748/d5f10a1a-5499-414d-8d8f-ec84ac608200">
<img width="479" alt="截屏2023-05-17 22 33 18" src="https://github.com/slint-ui/slint/assets/33011748/a8dba3fa-aa3c-4f00-851e-5569259c3ec3">

**Comment by ogoffart** [comment]
I could reproduce the problem.
And I measured with heap track,  looks like the curflpit is the femtovg text:

![image](https://github.com/slint-ui/slint/assets/959326/5be354ba-8599-4bc6-b8f4-98bc720d328e)

**Reply by ogoffart** [reply]
As @tronical pointed out to me, this is caused by https://github.com/slint-ui/slint/pull/2393/commits/8b7a744a51a722bd544cd806402108fc52838982 which increase the cache size to be a bit too large.
We probably should decrease these number to some more reasonable values
But the cache is still bounded so it won't grow forever.

**Comment by pcevent** [comment]
May I ask how to solve this problem?

**Comment by pcevent** [comment]
Okay, thank you for your reply.

**Comment by tronical** [comment]
Another workaround would be to use the Skia renderer. But we do need to fix this in either FemtoVG or our use of it. I'll create an issue in any case.
